### PR TITLE
add ==, isequal <, and isless for DataFrameRow and GroupKey

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2639.

As usual, before I add tests and update documentation please have a look at the implementation.

It would be good to have a decision on https://github.com/JuliaLang/julia/issues/40142 before finalizing this PR.

In particular we fix a bug according to which two `DataFrameRow` were `true` in `==` if they were the same row from the same data frame even if they contained `missing` (`missing` should be produced then).

### CODE REVIEW

**Comment by bkamins** [comment]
Ah - and it is mildly breaking as we now define `hash` for `DataFrameRow` and `GroupKey` to be consistent with respective `NamedTuple` hash.

This means in particular that this hash is different than `rowhash` used in grouping (we could make `rowhash` consistent with the current `hash` - @nalimilan - do you think it is worth doing?).

File: src/other/utils.jl
Comment (nalimilan) [inline src/other/utils.jl:132]: Is this faster than just doing `hash(NamedTuple(v), h)`?

File: src/dataframerow/dataframerow.jl
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl:517]:
```suggestion
        throw(ArgumentError("compared objects must have the same property " *
```

File: src/dataframerow/dataframerow.jl
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl:534]:
```suggestion
        throw(ArgumentError("compared objects must have the same property " *
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:562]: Maybe this duplication could be avoided by using an `@eval` loop? The only difference is the function which is used with `all`.

**Comment by nalimilan** [comment]
> This means in particular that this hash is different than `rowhash` used in grouping (we could make `rowhash` consistent with the new `hash` - @nalimilan - do you think it is worth doing?).

We probably don't care about this since it's internal, right?

**Comment by bkamins** [comment]
> We probably don't care about this since it's internal, right?

Yes - I have checked that we use rowhash only in one place in the code so it should be OK.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:562]: I will remove most of these codes anyway (and then I will look at how to reduce duplication).
However I am not super fan of `@eval` as then the stacktrace users receive when they make some error are very hard to understand for non-experts.

File: src/other/utils.jl
Comment (bkamins) [reply src/other/utils.jl:132]: doing `NamedTuple(v)` can have a prohibitive compilation cost for wide data frames. Also it is faster. Here are timings for 5-column data frame:
```
julia> @benchmark _nt_like_hash($df[1, :], UInt(0))
BenchmarkTools.Trial:
  memory estimate:  368 bytes
  allocs estimate:  14
  --------------
  minimum time:     589.655 ns (0.00% GC)
  median time:      598.276 ns (0.00% GC)
  mean time:        654.433 ns (1.49% GC)
  maximum time:     9.713 μs (89.57% GC)
  --------------
  samples:          10000
  evals/sample:     174

julia> @benchmark hash(NamedTuple($df[1, :]))
BenchmarkTools.Trial:
  memory estimate:  416 bytes
  allocs estimate:  11
  --------------
  minimum time:     1.310 μs (0.00% GC)
  median time:      1.450 μs (0.00% GC)
  mean time:        1.537 μs (0.00% GC)
  maximum time:     5.290 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10
```

**Comment by bkamins** [comment]
Also `rowhash` ignores column names for speed, so indeed they do not have to match.

File: src/dataframerow/dataframerow.jl
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl:479]: Maybe call these `eqfun` and `cmpfun` or something explicit like that?

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:570]: You defined a helper like this, the same definition would work for both types, right?
```julia
_getnames(x::GroupKey) = parent(x).cols 
_getnames(x::DataFrameRow) = _names(x)
```

File: src/dataframerow/dataframerow.jl
Comment (bkamins) [reply src/dataframerow/dataframerow.jl:479]: OK - will change.

File: src/groupeddataframe/groupeddataframe.jl
Comment (bkamins) [reply src/groupeddataframe/groupeddataframe.jl:570]: I will change this. I was thinking about it, but was too lazy ;(.

**Comment by bkamins** [comment]
OK - I hope I fixed everything :). This was hard (but hopefully now Julia Base will be also more consistent with `<` handling).

File: src/dataframerow/dataframerow.jl
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl:463]:
```suggestion
# this is required as == does not allow for comparison between tuples and vectors
```

File: test/dataframerow.jl
Comment (nalimilan) [inline test/dataframerow.jl:602]: It could be interesting to test the equivalent of `(1, missing) < (2, 1)`, which is a corner case that shouldn't give `missing` but is easy to get wrong.

Comment (bkamins) [reply test/dataframerow.jl:602]: I will add it

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
@pdeffebach - this is what we concluded is a good design. Could you please comment on it before merging (as you were against it)?

**Comment by pdeffebach** [comment]
Sounds good! I'm fine with this.

File: src/dataframerow/utils.jl
Comment (bkamins) [inline src/dataframerow/utils.jl:458]: it is interesting that we do not have this covered by tests. I will add some.

Comment (bkamins) [reply src/dataframerow/utils.jl]: it turns out we do not use these functions. I have removed them. @nalimilan - could you please have a look if we would ever need them? (I have never used them)

File: test/dataframerow.jl
Comment (bkamins) [inline test/dataframerow.jl:176]:
```suggestion
```

File: test/dataframerow.jl
Comment (bkamins) [inline test/dataframerow.jl:192]:
```suggestion
        @test hash(DataFrameRow(df, i, :), h) ==
              hash(gks[i], h) ==
              hash(NamedTuple(DataFrameRow(df, i, :)), h)
```

File: src/dataframerow/utils.jl
Comment (nalimilan) [reply src/dataframerow/utils.jl]: AFAICT one guy removed the last use of `findrow` and `group_rows` in a recent PR. :-D https://github.com/JuliaData/DataFrames.jl/pull/2641

It's weird that coverage didn't notice this. Maybe we only looked at the changed code and not at unrelated parts?

File: src/dataframerow/utils.jl
Comment (nalimilan) [inline src/dataframerow/utils.jl:450]: I think `RowGroupDict` can also be removed.

Comment (bkamins) [reply src/dataframerow/utils.jl:450]: Indeed - I am  running tests to double check. It is astonishing how much we have reworked internally in this release.

File: src/dataframerow/utils.jl
Comment (bkamins) [reply src/dataframerow/utils.jl]: There were tests that checked correctness of internal functions.

File: src/dataframerow/utils.jl
Comment (bkamins) [reply src/dataframerow/utils.jl:450]: OK - all seems clean. I will merge the PR after CI passes.

Comment (nalimilan) [reply src/dataframerow/utils.jl:450]: Now that this file contains only grouping code, we will be able to move it to the corresponding folder and rename it without touching anything else. :-)

Comment (bkamins) [reply src/dataframerow/utils.jl:450]: OK will move it in this PR as otherwise we will forget. Also `nonunique` uses it, but I think it is not a problem.

**Comment by bkamins** [comment]
Thank you!

**Comment by pdeffebach** [comment]
This allows 

```
(a = 1, b = 2) in eachrow(df)
```

right?

**Comment by bkamins** [comment]
Right, and even more e.g.:
```
julia> df = DataFrame(a=1:5, b=2:6)
5×2 DataFrame
 Row │ a      b     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      2
   2 │     2      3
   3 │     3      4
   4 │     4      5
   5 │     5      6

julia> (a=1, b=2) in eachrow(df)
true

julia> Ref((a=4, b=4)) .< eachrow(df)
5-element BitVector:
 0
 0
 0
 1
 1
```

**Pull request merged**
